File: app/src/main/java/com/left/gank/ui/daily/DailyGirlAdapter.java
package com.left.gank.ui.daily;

import com.left.gank.butterknife.adapter.BaseAdapter;
import android.ly.business.domain.Girl;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.left.gank.R;
import com.left.gank.butterknife.holder.BindHolder;
import com.left.gank.butterknife.item.ItemModel;
import com.left.gank.listener.ItemClick;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

/**
 * Create by LingYan on 2016-07-05
 */
public class DailyGirlAdapter extends BaseAdapter<DailyGirlAdapter.NormalHolder> {
    private ItemClick mMeiZiOnClick;
    private List<Girl> mGirlList = new ArrayList<>();

    DailyGirlAdapter() {
        setHasStableIds(true);
    }

    @Override
    public NormalHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new DailyGirlHolder(parent);
    }

    @Override
    public void onBindViewHolder(NormalHolder holder, int position) {

    }

    public void setOnItemClickListener(ItemClick onItemClickListener) {
        mMeiZiOnClick = onItemClickListener;
    }

    @Override
    public int getItemCount() {
        return mGirlList.size();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    void refillItem(List<Girl> girlList) {
        int size = mGirlList.size();
        mGirlList.clear();
        notifyItemRangeRemoved(0, size);
        appendItem(girlList);
    }

    public void appendItem(List<Girl> girlList) {
        mGirlList.addAll(girlList);
        notifyItemRangeInserted(getItemCount(), girlList.size());
    }

    @Override
    public void destroy() {

    }

    public static class DailyGirlHolder extends NormalHolder<NormalItem> implements View.OnClickListener {
        static final int LAYOUT = R.layout.adapter_daily_girl;

        @BindView(R.id.title)
        TextView txtTitle;

        Girl girl;

        public DailyGirlHolder(ViewGroup parent) {
            super(parent, LAYOUT);
        }

        @Override
        public void onClick(View v) {
        }

        @Override
        public void bindHolder(NormalItem item) {

        }
    }

    static class NormalItem extends ItemModel {

        @Override
        public int getViewType() {
            return 0;
        }
    }

    abstract static class NormalHolder<TT extends ItemModel> extends BindHolder<TT> {

        public NormalHolder(ViewGroup parent, int layoutRes) {
            super(parent, layoutRes);
        }

        void bindHolder(TT item, ItemCallback callback) {

        }
    }

    interface ItemCallback {
        void onItemClick(Girl girl);
    }
}

File: app/src/main/java/com/left/gank/ui/daily/DailyGirlAdapter.kt
package com.left.gank.ui.daily

import android.business.domain.Girl
import android.view.View
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.left.gank.R
import com.left.gank.butterknife.adapter.BaseAdapter
import com.left.gank.butterknife.adapter.BindHolder
import com.left.gank.butterknife.item.ItemModel
import kotlinx.android.synthetic.main.adapter_daily_girl.view.*

/**
 * Create by LingYan on 2016-07-05
 */
class DailyGirlAdapter internal constructor() : BaseAdapter<BindHolder<ItemModel>>() {
    private var itemClick: ItemClickListener? = null
    private val items = mutableListOf<ItemModel>()

    @Suppress("UNCHECKED_CAST")
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BindHolder<ItemModel> {
        return DailyGirlHolder(parent, itemClick) as BindHolder<ItemModel>
    }

    override fun onBindViewHolder(holder: BindHolder<ItemModel>, position: Int) {
        holder.bindHolder(items[position])
    }

    fun setOnItemClickListener(itemClick: ItemClickListener) {
        this.itemClick = itemClick
    }

    override fun getItemCount(): Int = items.size

    fun refillItem(girlList: List<Girl?>) {
        items.clear()
        appendItem(girlList)
    }

    fun appendItem(girlList: List<Girl?>) {
        girlList.filterNotNull().forEach {
            items.add(NormalItem(it))
        }
    }

    class DailyGirlHolder(parent: ViewGroup, private val call: ItemClickListener?) : BindHolder<NormalItem>(parent, R.layout.adapter_daily_girl) {

        override fun bindHolder(item: NormalItem) {
            val girl = item.girl
            Glide.with(itemView.context)
                    .load(girl.url)
                    .apply(RequestOptions.diskCacheStrategyOf(DiskCacheStrategy.ALL))
                    .into(itemView.image)
            itemView.setOnClickListener { call?.onItemClick(girl, itemView) }
        }
    }

    class NormalItem(val girl: Girl) : ItemModel() {

        override fun getViewType(): Int {
            return 0
        }
    }

    interface ItemClickListener {
        fun onItemClick(girl: Girl, view: View)
    }
}
